cpp/vector.hpp

```diff
@@ -0,0 +1,379 @@
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <initializer_list>
+#include <numeric>
+#include <type_traits>
+
+/**
+ * @brief 幾何学としてのベクトル
+ *
+ * @tparam T 型 省略したらdouble
+ * @tparam Dim 次元 省略したら2
+ */
+template <typename T = double, int Dim = 2, std::enable_if_t<std::is_scalar_v<T> && (Dim > 0)>* = nullptr>
+struct Vector {
+    std::array<T, Dim> v; //!< ベクトルの成分表示
+
+    /**
+     * @brief デフォルトコンストラクタ
+     * ゼロベクトル
+     */
+    constexpr Vector() noexcept : v() {}
+    /**
+     * @brief 可変長引数によるコンストラクタ
+     * 
+     * @tparam Args 
+     */
+    template <typename... Args, std::enable_if_t<(sizeof...(Args) == Dim)>* = nullptr>
+    constexpr Vector(Args... args) noexcept : v{args...} {}
+    
+    /**
+     * @brief i番目の成分を返す
+     * 
+     * @param i
+     * @return T& 
+     */
+    constexpr T& operator[](int i) noexcept {
+        return v[i];
+    }
+    /**
+     * @brief i番目の成分を返す
+     * 
+     * @param i
+     * @return T& 
+     */
+    constexpr const T& operator[](int i) const noexcept {
+        return v[i];
+    }
+
+    /**
+     * @brief 加算代入演算子
+     * 
+     * @param rhs 
+     * @return Vector<T, Dim>& 
+     */
+    constexpr Vector<T, Dim>& operator+=(const Vector<T, Dim>& rhs) noexcept {
```

Comment:
クラス定義の中の `Vector<T, Dim>` はすべて `Vector` と書けます。
これはエラーではないので、修正は必須ではないです。

```suggestion
    constexpr Vector& operator+=(const Vector& rhs) noexcept {
```

Reply:
たしかにそうなので省きます…
ありがとうございます